pkg/restore/cdclog/puller.go

```diff
@@ -0,0 +1,181 @@
+	storage storage.ExternalStorage) (*EventPuller, error) {
+	var (
+		ddlDecoder        *JSONEventBatchMixedDecoder
+		ddlFileIndex      int
+		rowChangedDecoder *JSONEventBatchMixedDecoder
+		rowFileIndex      int
+	)
+	if len(ddlFiles) == 0 {
+		log.Info("There is no ddl file to restore")
+	} else {
+		data, err := storage.Read(ctx, ddlFiles[0])
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		ddlFileIndex++
+		ddlDecoder, err = NewJSONEventBatchDecoder(data)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	if len(rowChangedFiles) == 0 {
+		log.Info("There is no row changed file to restore")
+	} else {
+		data, err := storage.Read(ctx, rowChangedFiles[0])
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		rowFileIndex++
+		rowChangedDecoder, err = NewJSONEventBatchDecoder(data)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	return &EventPuller{
+		schema: schema,
+		table:  table,
+
+		ddlDecoder:        ddlDecoder,
+		rowChangedDecoder: rowChangedDecoder,
+
+		ddlFiles:            ddlFiles,
+		rowChangedFiles:     rowChangedFiles,
+		ddlFileIndex:        ddlFileIndex,
+		rowChangedFileIndex: rowFileIndex,
+
+		storage: storage,
+	}, nil
+}
+
+// PullOneEvent pulls one event in ts order.
+// The Next event which can be DDL item or Row changed Item depends on next commit ts.
+func (e *EventPuller) PullOneEvent(ctx context.Context) (*SortItem, error) {
+	var err error
+	// ddl exists
+	if e.ddlDecoder != nil {
+		// current file end, read next file if next file exists
+		if !e.ddlDecoder.HasNext() && e.ddlFileIndex < len(e.ddlFiles) {
+			path := e.ddlFiles[e.ddlFileIndex]
+			data, err := e.storage.Read(ctx, path)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if len(data) > 0 {
+				e.ddlFileIndex++
+				e.ddlDecoder, err = NewJSONEventBatchDecoder(data)
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+			}
+		}
+		// set current DDL item first
+		if e.currentDDLItem == nil {
+			e.currentDDLItem, err = e.ddlDecoder.NextEvent(DDL)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+	}
+	// dml exists
+	if e.rowChangedDecoder != nil {
+		// current file end, read next file if next file exists
+		if !e.rowChangedDecoder.HasNext() && e.rowChangedFileIndex < len(e.rowChangedFiles) {
+			path := e.rowChangedFiles[e.rowChangedFileIndex]
+			data, err := e.storage.Read(ctx, path)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if len(data) != 0 {
+				e.rowChangedFileIndex++
+				e.rowChangedDecoder, err = NewJSONEventBatchDecoder(data)
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+			}
+		}
+		if e.currentRowChangedItem == nil {
+			e.currentRowChangedItem, err = e.rowChangedDecoder.NextEvent(RowChanged)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+	}
+
+	var returnItem *SortItem
+	switch {
+	case e.currentDDLItem != nil:
+		if e.currentDDLItem.LessThan(e.currentRowChangedItem) {
```

Comment:
is it true that `e.currentRowChangedItem` must be non-nil here?

Reply:
if it's nil, LessThan will return true.